### PR TITLE
Add support for MySQL Coordinator

### PR DESCRIPTION
**Pull request by heheh13**

### CODE REVIEW

File: pkg/controller/rbac.go [hunk 1]
```diff
@@ -149,8 +158,22 @@ func (c *Controller) getPolicyNames(db *api.MySQL) (string, error) {
 func (c *Controller) ensureDatabaseRBAC(db *api.MySQL) error {
```
Comment (spectro30) [inline pkg/controller/rbac.go:158]: take a look into the whole function

File: pkg/controller/router.go [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+			eventer.EventReasonSuccessful,
+			"succesfully %v statefulset",
+			vt,
+		)
+	}
+	return nil
+}
+func (c *Controller) ensureRouterConfigSecret(db *api.MySQL) error {
+	secretMeta := metav1.ObjectMeta{
+		Name:      "router-config",
```
Comment (spectro30) [inline pkg/controller/router.go:76]: use constant

File: pkg/controller/rbac.go [hunk 3]
```diff
@@ -149,8 +158,22 @@ func (c *Controller) getPolicyNames(db *api.MySQL) (string, error) {
...
 	saName := db.Spec.PodTemplate.Spec.ServiceAccountName
 	if saName == "" {
-		saName = db.OffshootName()
-		db.Spec.PodTemplate.Spec.ServiceAccountName = saName
+		db.Spec.PodTemplate.Spec.ServiceAccountName = db.OffshootName()
```
Comment (spectro30) [inline pkg/controller/rbac.go:161]: recheck the access

File: pkg/controller/rbac.go [hunk 4]
```diff
@@ -149,8 +158,22 @@ func (c *Controller) getPolicyNames(db *api.MySQL) (string, error) {
...
+				},
+			}
+		} else {
+
+			routerSaName := db.Spec.Topology.InnoDBCluster.Router.PodTemplate.Spec.ServiceAccountName
+			if routerSaName == "" {
+				db.Spec.Topology.InnoDBCluster.Router.PodTemplate.Spec.ServiceAccountName = db.OffshootName()
+			}
+		}
 	}
```
Comment (spectro30) [inline pkg/controller/rbac.go:177]: create new sa for router

File: pkg/controller/service.go [hunk 5]
```diff
@@ -129,10 +129,86 @@ func (c *Controller) ensureService(db *api.MySQL) error {
...
+		vt, err := c.ensureRouterPrimaryService(db)
+		if err != nil {
+			return err
+		}
+		if vt == kutil.VerbCreated {
+			c.Recorder.Event(
+				db,
+				core.EventTypeNormal,
+				eventer.EventReasonSuccessful,
+				"Successfully created service for Innodb Cluster ",
```
Comment (spectro30) [inline pkg/controller/service.go:142]: Successfully created service for `router name`

File: pkg/controller/service.go [hunk 6]
```diff
@@ -129,10 +129,86 @@ func (c *Controller) ensureService(db *api.MySQL) error {
...
+		in.Annotations = svcTemplate.Annotations
+
+		if db.Spec.Topology != nil && *db.Spec.Topology.Mode == api.InnoDBClusterModeGroup {
+
+			in.Spec.Ports = ofst.PatchServicePorts(
+				core_util.MergeServicePorts(in.Spec.Ports, []core.ServicePort{
+					{
+						Name: "read-write",
+						Port: 6446,
+						//TargetPort: intstr.FromString(api.MySQLDatabasePortName),
```
Comment (spectro30) [inline pkg/controller/service.go:181]: check if it produce any unwanted logs in operator

File: pkg/controller/service.go [hunk 7]
```diff
@@ -147,6 +223,7 @@ func (c *Controller) ensurePrimaryService(db *api.MySQL) (kutil.VerbType, error)
 		in.Annotations = meta_util.OverwriteKeys(in.Annotations, svcTemplate.Annotations)
 
 		in.Spec.Selector = db.OffshootSelectors()
+		in.Annotations = svcTemplate.Annotations
```
Comment (spectro30) [inline pkg/controller/service.go:226]: did the same thing on line 223

File: pkg/controller/health.go [hunk 8]
```diff
@@ -45,6 +45,36 @@ const (
 	TLSValueSkipVerify = "skip-verify"
 )
 
+// health checker algorithm
```
Comment (spectro30) [inline pkg/controller/health.go:48]: remove commented codes

File: pkg/controller/health.go [hunk 9]
```diff
@@ -154,66 +177,43 @@ func (c *Controller) CheckMySQLHealthOnce() {
...
-							Reason:             api.DatabaseAcceptingConnectionRequest,
-							ObservedGeneration: db.Generation,
-							Message:            fmt.Sprintf("The MySQL: %s/%s is accepting client requests.", db.Namespace, db.Name),
-						})
-					return db.UID, in
-				},
-				metav1.UpdateOptions{},
-			)
+			// Update "AcceptingConnection" to "true"
+			err = c.updateMySQLStatusConditions(ctx, db, kmapi.Condition{
```
Comment (spectro30) [inline pkg/controller/health.go:211]: maybe add this function in api machinery if it's not already there!

File: pkg/controller/health.go [hunk 10]
```diff
@@ -222,68 +222,133 @@ func (c *Controller) CheckMySQLHealthOnce() {
...
-								Status:             core.ConditionFalse,
-								Reason:             api.ReadinessCheckFailed,
-								ObservedGeneration: db.Generation,
-								Message:            fmt.Sprintf("The MySQL: %s/%s is not ready.", db.Namespace, db.Name),
-							})
-						return db.UID, in
-					},
-					metav1.UpdateOptions{},
-				)
+				err := c.updateMySQLStatusConditions(ctx, db, kmapi.Condition{
```
Comment (spectro30) [inline pkg/controller/health.go:269]: maybe add this function in api machinery if it's not already there!

File: pkg/controller/health.go [hunk 11]
```diff
@@ -369,14 +433,13 @@ func (c *Controller) getMySQLClient(ctx context.Context, db *api.MySQL, dns stri
 	}
 	tlsParam := ""
 	if db.Spec.TLS != nil {
-		serverSecret, err := c.Client.CoreV1().Secrets(db.Namespace).Get(ctx, db.MustCertSecretName(api.MySQLServerCert), metav1.GetOptions{})
+		serverSecret, err := c.Client.CoreV1().Secrets(db.Namespace).Get(ctx, db.MustCertSecretName(api.MySQLClientCert), metav1.GetOptions{})
```
Comment (spectro30) [inline pkg/controller/health.go:436]: shouldn't we use MySQLServerCert?

**Review by spectro30 — CHANGES_REQUESTED** [review]

File: pkg/controller/router.go [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kutil "kmodules.xyz/client-go"
+	kmapi "kmodules.xyz/client-go/api/v1"
+	app_util "kmodules.xyz/client-go/apps/v1"
+	core_util "kmodules.xyz/client-go/core/v1"
+	meta_util "kmodules.xyz/client-go/meta"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (heheh13) [reply pkg/controller/router.go:23]: will try next .

**Pull request merged**